### PR TITLE
feat: add hot start post

### PR DESCRIPTION
**Pull request by be-marc**

### CODE REVIEW

File: mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+# Scope
+
+Hotstarting a learner resumes the training from an already fitted model.
+For example, training an XGBoost model for additional 500 boosting iterations.
+Hotstarting in mlr3 works with a cache of already trained models.
+This cache is called a `r ref("HotstartStack")` which is automatically created while tuning.
+We distinguish between forward and backward hotstarting.
+We start this post with backward hotstarting and then talk about the less efficient forward hotstarting.
+
+# Backward Hotstarting
```
Comment (sebffischer) [inline mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd:40]: Also bevor Ich hier alles umstellst und es dir nicht gefällt: 

Ich würde die Reihenfolge anders machen. Ich würde zuerst zeigen wie man manuell hotstarted damit man einmal verstanden hat wie das geht. Dann würde ich motivieren, dass man da ja beim tunen verwenden kann, also einen 2D search space printen und andhand dessen herzeigen wie hotsarting genutzt werden kann (z.B. ausrechnen wie file boosting nrounds man mit und ohne hotstarting insgesamt braucht) und dann erst mit dem tunen kommen.

File: mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+```
+
+# Scope
+
+Hotstarting a learner resumes the training from an already fitted model.
+For example, training an XGBoost model for additional 500 boosting iterations.
+Hotstarting in mlr3 works with a cache of already trained models.
+This cache is called a `r ref("HotstartStack")` which is automatically created while tuning.
+We distinguish between forward and backward hotstarting.
+We start this post with backward hotstarting and then talk about the less efficient forward hotstarting.
```
[the remark this answers is not in the document]
Comment (sebffischer) [reply mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd:38]: außerdem würde ich das nicht so generisch sagen. Also hier wirkt es so als wäre FORWARD besser als BACKWARD, aber es kommt halt insgesamt einfach stark auf den learner und die daten etc. an ob sich was lohnt...

File: mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+We activate hotstarting with the `allow_hotstart` option.
+When running a grid search with hotstarting, the grid is sorted by the hot start parameter.
+This means the models with 2000 trees are trained first.
+The models with less than 2000 trees hot start on the 2000 trees models which allows the training to be completed immediately.
+
+```{r}
+#| eval: false
+
+instance = tune(
```
Comment (sebffischer) [inline mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd:71]: indentiation

File: mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+  resampling = rsmp("holdout"),
+  measure = msr("classif.ce"),
+  allow_hotstart = TRUE
+)
+```
+
+For comparison, we perform the same tuning without hotstarting.
+
+```{r}
+#| eval: false
```
Comment (sebffischer) [inline mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd:84]: indentiation

File: mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+  geom_point(aes(color = id)) +
+  scale_colour_viridis_d(name = "hotstarting", labels = c("Activated", "Deactivated"), end = 0.8) +
+  ylab("Batch") +
+  xlab("Time of Completion [seconds]") +
+  theme_minimal()
+```
+
+# Forward Hotstarting
+
+Forward hotstarting is currently only supported by XGBoost.
```
Comment (sebffischer) [inline mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd:128]:
```suggestion
Forward hotstarting is currently only supported by XGBoost and LightGBM.
```

File: mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd (hunk 2 above)
[the remark this answers is not in the document]
Comment (sebffischer) [reply mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd:38]: das sagst du ja später eh aber würde es trotzdem hier rausnehmen

File: mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+However, we have observed that hotstarting only provides a speed advantage for very large datasets and models with more than 5000 boosting rounds.
+The reason is that copying the models from the main process to the workers is a major bottleneck.
+The parallelization package future copies the models sequentially to the workers.
+Consequently, it takes a long time until the last worker can even start.
+Moreover, copying itself consumes a lot of time, and copying the model back from the worker blocks the main process again.
+During the development process, we overestimated the speed benefits of hotstarting and underestimated the overhead of parallelization.
+We can therefore only advise against using forward hotstarting during tuning.
+It is much more efficient to use the internal early-stopping mechanism of XGBoost.
```
Comment (sebffischer) [inline mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd:136]:
```suggestion
It is much more efficient to use the internal early-stopping mechanism of XGBoost instead of tuning the number of boosting iterations.
```

File: mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+
+Nevertheless, forward hotstarting can be useful without parallelization.
+If you have an already trained model and want to add more boosting iteration to it.
+In this example, the `learner_5000` is the already trained model.
+We create a new learner with the same hyperparameters but double the number of boosting iteration.
+To activate hotstarting, we create a `HotstartStack` and copy it to the `$hotstart_stack` slot of the new learner.
+
+```{r}
+#| eval: false
+
```
Comment (sebffischer) [inline mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd:153]: Das hat nix so richtig mit dem Post zu tun, aber wieso kann die `$train()` Methode nicht einfach ein `self$hostart_stack = HostartStack$new(learner)` machen falls es schon ein Modell gibt, das sich zum hostarten eignet?

File: mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+  xlab("Time of Completion [seconds]") +
+  theme_minimal()
+```
+
+# Forward Hotstarting
+
+Forward hotstarting is currently only supported by XGBoost.
+However, we have observed that hotstarting only provides a speed advantage for very large datasets and models with more than 5000 boosting rounds.
+The reason is that copying the models from the main process to the workers is a major bottleneck.
+The parallelization package future copies the models sequentially to the workers.
```
Comment (sebffischer) [inline mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd:131]: Also diese Details sollten finde ich in eine foldable callout box (siehe hier https://quarto.org/docs/authoring/callouts.html), weil für die meisten ist das zu technisch.

File: mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd [hunk 9]
```diff
@@ -0,0 +1,191 @@
...
+Consequently, it takes a long time until the last worker can even start.
+Moreover, copying itself consumes a lot of time, and copying the model back from the worker blocks the main process again.
+During the development process, we overestimated the speed benefits of hotstarting and underestimated the overhead of parallelization.
+We can therefore only advise against using forward hotstarting during tuning.
```
Comment (sebffischer) [inline mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd:135]:
```suggestion
We can therefore only advise to be cautious and to cover the runtime when relying on early stopping during tuning. 
```

File: mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+It is much more efficient to use the internal early-stopping mechanism of XGBoost.
+This eliminates the need to copy models to the worker.
+See the [gallery post](/gallery/optimization/2022-11-04-early-stopping-with-xgboost) on early stopping for an example.
+We might fix forward hotstarting in the future with workers that read hotstart models from a database in parallel.
```
Comment (sebffischer) [inline mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd:139]:
```suggestion
We might improve the efficiency of the hotstarting mechanism in the future, if there are convincing use cases.
```

File: mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+However, the need for this seems to be rather small.
```
Comment (sebffischer) [inline mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd:140]:
```suggestion
```

File: mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+Models that can be trained iteratively usually have a very efficient internal early-stopping mechanism.
```
Comment (sebffischer) [inline mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd:141]:
```suggestion
```

File: mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd (hunk 7 above)
Comment (sebffischer) [reply mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd:153]: und wie oben schon gesagt, ich finde das muss als erstes kommen.

Comment (sebffischer) [reply mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd:153]: ich würde vielleicht sogar noch zusätzlich zeigen was passiert wenn der hotstart stack 2 Modelle hat, dass eben dann das gewählt wird was am nächsten dran ist bezüglich des hotstarting parameters. 

Außerdem muss hier noch erwähnt werden, dass alle anderen Parameter dann gleich sein müssen.

File: mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+```{r}
+learner_10000$state$train_time - learner_5000$state$train_time
+```
+
+Training the model from the beginning would have taken about two minutes.
+This means, without parallelization, we get the expected speed advantage.
+
+# Conclusion
+
+Backward hotstarting significantly reduces the time to tune a random forest.
```
Comment (sebffischer) [inline mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd:189]:
```suggestion
We have seen how mlr3 enables to reduce the training time, by building on a hotstart stack of already trained learners.
```

File: mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd [hunk 14]
```diff
@@ -0,0 +1,191 @@
...
+Whereas forward hotstarting is less useful due to the parallelization overhead and the very efficient internal early stopping mechanisms.
```
Comment (sebffischer) [inline mlr-org/gallery/optimization/2022-12-09-hotstart/index.qmd:190]:
```suggestion
One has to be careful, however, when using forward hotstarting during tuning because of the high parallelization overhead that arises from copying the models between the processes. 
If a model has an internal early stopping implementation, it should usually be relied upon instead of using the mlr3 hotstarting mechanism.
```
